### PR TITLE
rpc: backpressured RPC server (bump jsonrpsee 0.20)

### PR DESCRIPTION
**Pull request by niklasad1**
This is a rather big change in jsonrpsee, the major things in this bump are:
   - Server backpressure (the subscription impls are modified to deal with that)
   - Allow custom error types / return types (remove jsonrpsee::core::Error and jsonrpee::core::CallError)
   - Bug fixes (graceful shutdown in particular not used by substrate anyway)
   - Less dependencies for the clients in particular
   - Return type requires Clone in method call responses
   - Moved to tokio channels
   - Async subscription API (not used in this PR)

Major changes in this PR:
- The subscriptions are now bounded and if subscription can't keep up with the server it is dropped
- CLI: add parameter to configure the jsonrpc server bounded message buffer (default is 64)
- Add our own subscription helper to deal with the unbounded streams in substrate

The most important things in this PR to review is the added helpers functions in `substrate/client/rpc/src/utils.rs` and the rest is pretty much chore.

Regarding the "bounded buffer limit" it may cause the server to handle the JSON-RPC calls 
slower than before.

The message size limit is bounded by "--rpc-response-size" thus "by default 10MB * 64 = 640MB" 
but the subscription message size is not covered by this limit and could be capped as well.

Hopefully the last release prior to 1.0, sorry in advance for a big PR

Previous attempt: https://github.com/paritytech/substrate/pull/13992

Resolves https://github.com/paritytech/polkadot-sdk/issues/748, resolves https://github.com/paritytech/polkadot-sdk/issues/627

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3501242

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3501532

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3502392

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3727955

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3759426

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3759409

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3759397

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3760414

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3764623

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3777262

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3779909

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3784467

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3788325

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3789247

**Comment by nazar-pc** [comment]
I'm following PR to see when it lands and I find it annoying how many notifications about force pushes just to fix clippy are happening. Can't you run clippy locally and fix it all at once before pushing to GitHub in the first place?

**Comment by niklasad1** [comment]
Sorry about that but I'm force pushing to make it easy to backport to a previous polkadot release, I haven't fixed the tests when I rebased/merged it to master.

Would be nice if Github could just subscribe you when it's actually PR and not draft :pray:

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3791190

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3831921

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3927733

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3955065

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3958892

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4019206

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4086869

File: substrate/client/service/src/lib.rs
Comment (niklasad1) [inline substrate/client/service/src/lib.rs:111]: breaking change: maybe change this to futures::Stream?

File: substrate/client/rpc-servers/Cargo.toml
Comment (niklasad1) [inline substrate/client/rpc-servers/Cargo.toml:22]: "util" added for `option_layer` API.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4124607

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4127163

File: substrate/client/rpc-servers/src/lib.rs
Comment (niklasad1) [inline substrate/client/rpc-servers/src/lib.rs:91]: Changed this to "bind" the port before starting the server in order to 
enable "host filter" for the actual port.

File: substrate/client/rpc-servers/src/lib.rs
Comment (niklasad1) [inline substrate/client/rpc-servers/src/lib.rs:142]: added IPv6 loopback as well

File: substrate/client/rpc-spec-v2/src/chain_head/chain_head.rs
Comment (niklasad1) [inline substrate/client/rpc-spec-v2/src/chain_head/chain_head.rs:351]:
```suggestion
			.spawn_blocking("substrate-rpc-subscription", Some("rpc"), fut.boxed());
```

File: substrate/client/rpc-spec-v2/src/transaction/transaction.rs
Comment (niklasad1) [inline substrate/client/rpc-spec-v2/src/transaction/transaction.rs:102]: reject is async now and this is moved to the async block for that reason.

File: substrate/client/rpc/src/lib.rs
Comment (niklasad1) [inline substrate/client/rpc/src/lib.rs:93]: This was provided by jsonrpsee before and I added something similar as a starting point.

The new rpc-spec has specified how to deal with if the server can't keep with the number of messages.
For now each implementation can specify how many item is allowed by kept in memory until a subscription is dropped and is this PR I used 16 but if anyone wants another number for a specific subscription it's doable as well.

This is needed because for instance import_streams and storage_streams are unbounded.

//cc @koute maybe you have some ideas for this until we abandon the old rpc APIs? :)

File: substrate/client/rpc/src/state/tests.rs
Comment (niklasad1) [inline substrate/client/rpc/src/state/tests.rs:284]: This was changed when jsonrpsee moved to tokio mpsc

File: substrate/client/rpc/src/state/tests.rs
Comment (niklasad1) [inline substrate/client/rpc/src/state/tests.rs:549]: This was changed when jsonrpsee moved to tokio mpsc

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4138222

File: substrate/client/rpc/src/lib.rs
Comment (niklasad1) [inline substrate/client/rpc/src/lib.rs:109]: Ideally, we would only initialize the stream after the subscription was "accepted" but
that would be a breaking change and I don't want to break it here.

Once the subscription has been "accepted" we can't "reject" except sending a special subscription notification for it.

File: substrate/client/rpc-spec-v2/src/chain_head/chain_head.rs
Comment (niklasad1) [inline substrate/client/rpc-spec-v2/src/chain_head/chain_head.rs:219]:
```suggestion
			.spawn("substrate-rpc-subscription", Some("rpc"), fut.boxed());
```

File: substrate/client/rpc/src/testing.rs
Comment (lexnv) [inline substrate/client/rpc/src/testing.rs:27]: nit: I do remember running into something similar in another PR and needing a custom executor, but can't quite remember. Was `sp_core::TaskExecutor` replaced by tokio to play nice with channels?

Comment (niklasad1) [reply substrate/client/rpc/src/testing.rs:27]: It's not needed anymore but before we use `tokio::mpsc::send_timeout` which only works on the tokio runtime.

I can revert it but I was too lazy to do it.

File: substrate/client/rpc/src/state/state_full.rs
Comment (lexnv) [inline substrate/client/rpc/src/state/state_full.rs:419]: Does this change the behavior of the `subscribe_storage`?
We previously rejected the subscription on `self.client.storage_changes_notification_stream` failures. Was this added as an optimization here that would have been otherwise reported by `storage_changes_notification_stream`?

File: substrate/client/rpc/src/state/state_full.rs
Comment (lexnv) [inline substrate/client/rpc/src/state/state_full.rs:460]: This is more clear to me, since its also explicit that we accept the subscription (something that was done under the hood by the jsonrpsee before).

I would probably add `rpc/lib/utils const DEFAULT_PENDING_BUFFER =16`, since that might change after more data is available to us.
Another approach here would be removing the capacity from `spawn_subscription_task`, then adding another `spawn_subscription_task_with_capacity` for more explicit use-cases. Leaving this up to you, looks good as it already

File: substrate/client/rpc/src/state/state_full.rs
Comment (lexnv) [reply substrate/client/rpc/src/state/state_full.rs:419]: To answer this: `state/mod.rs` did the same thing, I believe this is entirely fine

File: substrate/client/rpc/src/testing.rs
Comment (lexnv) [reply substrate/client/rpc/src/testing.rs:27]: That should be fine, no need to revert it imo. Thanks for the answer!

File: substrate/client/consensus/common/src/block_import.rs
Comment (lexnv) [inline substrate/client/consensus/common/src/block_import.rs:46]: dq: Does this need the `Clone` here? Wouldn't the `ImportResult` and the rest also need it?

Comment (niklasad1) [reply substrate/client/consensus/common/src/block_import.rs:46]: The clone is needed because it's part of result of a RPC method call.

Thus, no `ImportResult` is not part of that and no Clone needed.

**Review by lexnv — APPROVED** [review]
LGTM!

File: substrate/client/rpc/src/state/state_full.rs
Comment (niklasad1) [reply substrate/client/rpc/src/state/state_full.rs:460]: Deal but let's wait for other review comments and whether it makes sense to increase or decrease this for certain subscriptions.

**Comment by kayabaNerve** [comment]
Is there a reason this has *apparently* been stalled for a month and a half? I'd guess it's fallen through the cracks, yet I want to check there isn't simply a discussion which I missed.

**Comment by niklasad1** [comment]
> Is there a reason this has apparently been stalled for a month and a half? I'd guess it's fallen through the cracks, yet I want to check there isn't simply a discussion which I missed.

It's simply needs to be reviewed but since several people are off on Christmas holiday I'll hold off until the "next year" to get it in but nothing really that blocks it.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4840873

File: substrate/client/rpc/src/lib.rs
Comment (jsdw) [reply substrate/client/rpc/src/lib.rs:109]: That would def be an improvement I feel; something to try after this merges perhaps?

File: substrate/client/rpc/src/lib.rs
Comment (jsdw) [inline substrate/client/rpc/src/lib.rs:142]: Logic here looks solid to me; nice one!

File: substrate/client/rpc/src/state/state_full.rs
Comment (jsdw) [reply substrate/client/rpc/src/state/state_full.rs:460]: Did we gain any insight from burnins about whether any connections were being swiftly closed because of this buffer size?

Esp because we buffer while waiting for the client to accept, it feels not unlikely that some subscriptions might immediately hit this (thinking eg chainHead_follow, or possibly storage subscriptions or something)

Comment (jsdw) [reply substrate/client/rpc/src/state/state_full.rs:460]: But I agree that we could have a constant for the "default size" and then in a few cases we may use a bigger value.

File: substrate/client/rpc/src/lib.rs
Comment (jsdw) [inline substrate/client/rpc/src/lib.rs:50]: Nit; could be moved to a utils.rs file

Comment (jsdw) [reply substrate/client/rpc/src/lib.rs:50]: Also; we could add a couple of tests for this pipe_from_stream behaviour?

File: substrate/client/rpc-api/src/author/error.rs
Comment (jsdw) [inline substrate/client/rpc-api/src/author/error.rs:87]: How do we ensure not to use the same error code twice? Eg `BASE_ERROR + 2` comes up in a separate file.

Comment (jsdw) [reply substrate/client/rpc-api/src/author/error.rs:87]: Aah, It's a different BASE_ERROR!

**Review by jsdw — APPROVED** [review]
Great job @niklasad1! Some small comments/nits but things look solid overall, and it feels like a definite improvement! If the burnins have been looking good then I'm looking forward to seeing this merge!

File: substrate/client/rpc-api/src/author/error.rs
Comment (niklasad1) [reply substrate/client/rpc-api/src/author/error.rs:87]: Your comment makes sense anyway, globally there is nothing that checks that each error gets unique error code.

It's a manual thingy ^

File: substrate/client/rpc/src/state/state_full.rs
Comment (niklasad1) [reply substrate/client/rpc/src/state/state_full.rs:460]: > Did we gain any insight from burnins about whether any connections were being swiftly closed because of this buffer size?

Some where closed but no real data or anything to draw any conclusions on it but looked good to me but
it's been a while so I might have forgotten.

> Esp because we buffer while waiting for the client to accept, it feels not unlikely that some subscriptions might immediately hit this (thinking eg chainHead_follow, or possibly storage subscriptions or something)

We had so little traffic on the new rpc methods and pretty much the behavior only concern the old rpc subscriptions (storage subscription in particular)

File: substrate/client/rpc/src/lib.rs
Comment (niklasad1) [reply substrate/client/rpc/src/lib.rs]: Done and done

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4976797

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4976950

**Comment by niklasad1** [comment]
@paritytech/sdk-node 

Can you please have a look at this PR?
It's been opened for long time now and I would want to merge it but I want ensure that you are fine with it :pray:

**Comment by kayabaNerve** [comment]
@niklasad1 That failed to ping anyone from my point of view (which may be because that team isn't publicly viewable or may be because you had a typo. Apologies for the pointless comment if the former).

**Comment by niklasad1** [comment]
@kayabaNerve thanks, I changed to ping now :)

File: substrate/client/rpc/src/utils.rs
Comment (skunert) [inline substrate/client/rpc/src/utils.rs:59]: What is meant by configurable here? The BoundedVec has hardcoded `DEFAULT_BUF_SIZE`, no?

File: substrate/client/rpc/src/utils.rs
Comment (skunert) [inline substrate/client/rpc/src/utils.rs:117]: Nit: Probably just a matter of preference, but one of the tokio or futures select macros might make this a bit easier to read since you don't need to nest then.

File: substrate/client/service/src/lib.rs
Comment (skunert) [inline substrate/client/service/src/lib.rs:114]: Why are we passing `tokio::sync::Semaphore::MAX_PERMITS` as channel size here?

**Review by skunert — APPROVED** [review]
Only some nits/questions, overall LGTM!

File: substrate/client/rpc/src/utils.rs
Comment (niklasad1) [reply substrate/client/rpc/src/utils.rs]: Good catch, it was configurable before but I removed it after some review comments.

The rpc-v2-spec specifies how big the buffer should be and my hunch is that different subscriptions may need different buffer limits. I hope we could remove this helper function eventually and have specific logic per subscription.

File: substrate/client/rpc/src/utils.rs
Comment (niklasad1) [reply substrate/client/rpc/src/utils.rs:118]: Yepp, in general agree but tokio::select is easier to read but also a footgun if the future is not cancel-safe.

That's not the case here and I implemented a slightly weird borrow workaround to be able to not drop `next_fut` if another select branch is taken i.e, if pending send operation is pushed to `next_fut` we want not to drop it :)

File: substrate/client/service/src/lib.rs
Comment (niklasad1) [reply substrate/client/service/src/lib.rs:114]: Ideally we want to use `usize::MAX` but because `tokio::sync::mpsc::channel` panics then I changed it to `MAX_PERMITS` as upperbound and added a comment about it in the code.

**Pull request merged**
